### ISSUE TITLE
Testcontainers dockerhub rate limit in the CI

### ISSUE THREAD

**Issue by alafanechere** [airbytehq/airbyte#32418]
## Context
Our java connectors are using test containers to spawn test database with docker.
To run these tests in `airbyte-ci` we’re using a docker-in-docker pattern:
* A docker engine container (docker) is started with dagger, as a service
* The gradle container running the tests is bound to `dockerd` 

## Problem:
* We face transient rate limiting errors on DockerHub image pull.
* Tests use a lot of different images (Postgres, Mysql, mssql etc.). Tags used for these images are “dynamic”, e.g anytime a new Postgres version is pushed to DockerHub we might use a tag whose underneath image is updated, which leads to a pull.

## What we tried so far:
* We authenticated the `dockerd` container to DockerHub using `docker login`. This should give us 5000 pulls per day. However, it seems insufficient as the error continues to occur.
* We cache `/var/lib/docker` in a distributed cache volume. It allows us to cache re-used images. But a pull can happen anytime BuildKit evicts layers.
* We tried to retrieve the current consumption of our pull requests with DockerHub API. But their API does not return this metric an authenticated calls…

## Potential solution:
1. Give more disk space to make the Buildkit layer eviction less frequent. But we don’t have any proof ATM that eager pulls happen due to cache eviction.
2. Host docker proxy with a real addressable URL and make our `dockerd` engine connect to it
3. Use a third party docker hub proxy vendor  and make our `dockerd` engine connect to it
4. Upgrade to Dagger 0.9.0, which has new networking features, to allow our `dockerd` service container to reach the cluster level docker hub proxy.

## Code reference
The creation of the `dockerd` containers happens https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/pipelines/dagger/actions/system/docker.py#L16.

**Comment by postamar** [comment]
Thanks for filing this issue. It's particularly maddening that we have no insight into how this image pull quota is being used. What are we pulling? When? Where? The dockerhub API is particularly miserable in this respect.

Although we have no useful metrics it is a fact that this is disruptive to _all_ connector engineers and has been so during most PST business hours for the last 3 weeks at least. You're not hearing any noise about it because these engineers have been thoroughly inured to false negatives from the CI but it's real. It's also hard to have an alertable metric about this because:
- we're pulling from a server that we don't control;
- because the pull is being performed from within a test the result is a test failure, but in PRs tests routinely fail all the time for legitimate reasons;
- the nightlies are always red anyway and perhaps are less affected because they run at a time of day when the daily pull quota isn't yet exhausted.

We just don't have the infrastructure to alert on this right now. Presumably, we'd need to publish all of our testing logs into splunk or sumologic or whatever and set up an alert on the string "pull rate limit exceeded" or whatever. We've been groping around in the dark trying to fix this for weeks with no actionable signal other than our test runs stop failing sometimes. Worst of all, we may legitimately be pulling at least 5000 images per day, but we can't even tell! It's not that far-fetched that that could be the case:
- in source-postgres, until recently, we tested against more than 15 different images, I expect other connectors to be similar;
- we have many connectors which get built at least once per day;
- these images change over time, they're rebuilt to include security updates and so forth, in the case of postgres:* images they change every couple of months or so;
- the docker-in-docker cache volume is persisted on a best-effort basis but may be evicted from time to time although we've done all we can to maximize the persistence;
- we need to pull them for each runner, presumably, who knows if the dagger magiccache does its thing, we don't control it.

tl;dr we're in deep doodoo right now, please help us whichever way you feel is best

**Comment by alafanechere** [comment]
> Host docker proxy with a real addressable URL and make our dockerd engine connect to it

Without any change on the proxy side, I tried to `ping` it from our `dockerd` container:
```
exec sh -c set -o xtrace && ping -c 4 -q 172.20.83.84 > /dev/null || echo 'Host unreachable' && exit 1
```
Logs:
```
Host unreachable
```

Conclusion: I confirm that the `dockerd` container network does not have access to the kube service docker proxy IP

**Comment by cpdeethree** [comment]
2 is probably our current preferred solution. We will explore the best way to achieve this as there are a couple of different approaches.

**Comment by postamar** [comment]
@stephane-airbyte just hit the docker pull rate limit today. Any updates?

**Comment by cpdeethree** [comment]
Moved this to interrupt since it appears that this is more of a dumpster fire than originally thought.

Submitted a hacky workaround as a demo first: https://github.com/airbytehq/airbyte/pull/32962#issuecomment-1832766338

Then an actual solution is now in the works: https://github.com/airbytehq/airbyte/pull/32965

**Comment by cpdeethree** [comment]
Made `TAILSCALE_AUTH_KEY` secret available to `airbyte` repository in github

**Comment by git-phu** [comment]
solution is mainlined

connector ops reports it is working now

**Comment by git-phu** [comment]
note: this now means if you are connected to tailscale, you can use the pull through cache that exists on the ci cluster
